### PR TITLE
PMM-6450 new mongodb exporter

### PR DESCRIPTION
**Pull request by percona-csalguero**
Config to start the new MongoDB exporter

FB: https://github.com/Percona-Lab/pmm-submodules/pull/1074

### CODE REVIEW

**Comment by it-percona** [comment]
[CLA assistant check](https://cla-assistant.percona.com:443/percona/pmm-managed?pullRequest=441) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.percona.com:443/percona/pmm-managed?pullRequest=441) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.percona.com:443/check/percona/pmm-managed?pullRequest=441) it.</sub>

File: services/agents/mongodb.go [hunk 1]
```diff
@@ -25,11 +25,23 @@ import (
...
 	"github.com/percona/pmm/api/inventorypb"
 
+	"github.com/percona/pmm/version"
+
 	"github.com/percona/pmm-managed/models"
 )
 
 // mongodbExporterConfig returns desired configuration of mongodb_exporter process.
-func mongodbExporterConfig(service *models.Service, exporter *models.Agent, redactMode redactMode) *agentpb.SetStateRequest_AgentProcess {
+func mongodbExporterConfig(service *models.Service, exporter *models.Agent, redactMode redactMode, v *version.Parsed) *agentpb.SetStateRequest_AgentProcess {
```
Comment (BupycHuk) [inline services/agents/mongodb.go:34]:
```suggestion
func mongodbExporterConfig(service *models.Service, exporter *models.Agent, redactMode redactMode, pmmAgentVersion *version.Parsed) *agentpb.SetStateRequest_AgentProcess {
```

Comment (BupycHuk) [reply services/agents/mongodb.go:34]: should we pass *version.Parsed or just a string?

File: services/agents/mongodb.go [hunk 2]
```diff
@@ -25,11 +25,23 @@ import (
...
+	// New MongoDB Exporter will be released with PMM agent v2.10.0
+	pmmNewVersion, _ := version.Parse("2.10.0")
```
Comment (BupycHuk) [inline services/agents/mongodb.go:36]: I think it should be called somehow else.

File: services/agents/mongodb.go [hunk 3]
```diff
@@ -25,11 +25,23 @@ import (
...
+	if v.Less(pmmNewVersion) {
+		return mongodbExporterV1Config(service, exporter, redactMode)
```
Comment (BupycHuk) [inline services/agents/mongodb.go:38]: looks like the difference between these two function only in args and env variables, maybe it's better to return them, but keep common part in this function. What do you think?

File: services/agents/mongodb_test.go [hunk 4]
```diff
@@ -22,13 +22,16 @@ import (
...
 	"github.com/percona/pmm/api/inventorypb"
+	"github.com/percona/pmm/version"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
 	"github.com/percona/pmm-managed/models"
 )
 
 func TestMongodbExporterConfig(t *testing.T) {
+	pmmAgentVersion, _ := version.Parse("0.9.9")
```
Comment (BupycHuk) [inline services/agents/mongodb_test.go:33]: We don't have pmm-agent version lower then 2.0.0

File: services/agents/registry.go [hunk 5]
```diff
@@ -408,6 +408,17 @@ func (r *Registry) SendSetStateRequest(ctx context.Context, pmmAgentID string) {
...
 	}
 
+	pmmAgent, err := models.FindAgentByID(r.db.Querier, pmmAgentID)
+	if err != nil {
+		l.Errorf("Failed to get PMM Agent: %s.", err)
+		return
+	}
+	pmmAgentVersion, err := version.Parse(*pmmAgent.Version)
+	if err != nil {
+		l.Errorf("Failed to collect agents: %s.", err)
```
Comment (BupycHuk) [inline services/agents/registry.go:418]: Could you check when we added setting version into pmm-agent? If it was done before PMM-2.0 then it's ok.

**Review by BupycHuk — CHANGES_REQUESTED** [review]

File: services/agents/mongodb.go (hunk 1 above)
Comment (percona-csalguero) [reply services/agents/mongodb.go:34]: if we send a string, we must validate the parsing result inside the function and change the signature to also return the "possible" error. That's why I am passing the already parsed version

**Comment by BupycHuk** [comment]
@percona-csalguero don't forget to sign CLA

File: services/agents/registry.go (hunk 5 above)
Comment (percona-csalguero) [reply services/agents/registry.go:418]: It was added in PMM-2.0. It doesn't exist in master branch. 
Shouldn't we check for errors anyway?

File: services/agents/mongodb.go [hunk 6]
```diff
@@ -25,25 +25,55 @@ import (
 	"github.com/percona/pmm/api/agentpb"
 	"github.com/percona/pmm/api/inventorypb"
```
Comment (BupycHuk) [inline services/agents/mongodb.go:27]: please remove this blank line

File: services/agents/mongodb.go [hunk 7]
```diff
@@ -25,25 +25,55 @@ import (
...
 )
 
+var (
+	// New MongoDB Exporter will be released with PMM agent v2.10.0
+	pmmNewVersion, _ = version.Parse("2.10.0")
+)
+
 // mongodbExporterConfig returns desired configuration of mongodb_exporter process.
-func mongodbExporterConfig(service *models.Service, exporter *models.Agent, redactMode redactMode) *agentpb.SetStateRequest_AgentProcess {
+func mongodbExporterConfig(service *models.Service, exporter *models.Agent, redactMode redactMode, v *version.Parsed) *agentpb.SetStateRequest_AgentProcess {
```
Comment (BupycHuk) [inline services/agents/mongodb.go:39]: There was suggestion to rename `v` to `pmmAgentVersion`

File: services/agents/mongodb.go [hunk 8]
```diff
@@ -25,25 +25,55 @@ import (
...
 		pointer.GetString(service.Address),
 		pointer.GetString(exporter.Username),
 		pointer.GetString(exporter.Password),
 		pointer.GetString(exporter.MetricsPath),
 	)
 
+	var args, env []string
+	switch {
+	case v.Less(pmmNewVersion):
+		args = mongodbExporterV1ConfigArgs(tdp, exporter)
```
Comment (BupycHuk) [inline services/agents/mongodb.go:50]: we can have one function with two variables as result.

File: services/agents/mongodb_test.go [hunk 9]
```diff
@@ -65,13 +68,57 @@ func TestMongodbExporterConfig(t *testing.T) {
...
 	t.Run("EmptyUsername", func(t *testing.T) {
 		exporter.Username = nil
-		actual := mongodbExporterConfig(mongodb, exporter, exposeSecrets)
+		actual := mongodbExporterConfig(mongodb, exporter, exposeSecrets, pmmAgentVersion)
 		assert.Equal(t, "MONGODB_URI=mongodb://1.2.3.4:27017/?connectTimeoutMS=1000", actual.Env[0])
 	})
 }
+
+func TestNewMongodbExporterConfig(t *testing.T) {
+	pmmAgentVersion, _ := version.Parse("3.9.9")
```
Comment (BupycHuk) [inline services/agents/mongodb_test.go:83]: We don't have 3.9.9 too, correct versions are between 2.0.0 and 2.10.0

File: services/agents/registry.go (hunk 5 above)
Comment (BupycHuk) [reply services/agents/registry.go:418]: I mean which version of PMM-2.0? before september of 2019 or after?

File: services/platform/platform_test.go [hunk 10]
```diff
@@ -34,6 +34,7 @@ import (
 const devAuthHost = "check-dev.percona.com:443"
 
 func TestPlatformService(t *testing.T) {
+	t.Skip("Not working on travis ci")
```
Comment (BupycHuk) [inline services/platform/platform_test.go:37]: please pull latest changes from PMM-2.0, it should work now.

File: services/agents/registry.go [hunk 11]
```diff
@@ -408,6 +409,13 @@ func (r *Registry) SendSetStateRequest(ctx context.Context, pmmAgentID string) {
 		return
 	}
 
+	pmmAgent, err := models.FindAgentByID(r.db.Querier, pmmAgentID)
+	if err != nil {
+		l.Errorf("Failed to get PMM Agent: %s.", err)
+		return
+	}
+	pmmAgentVersion := validators.MustParseVersion(*pmmAgent.Version)
```
Comment (BupycHuk) [inline services/agents/registry.go:417]: looks like we should use `MustParseVersion` in other places, but not here.

**Review by BupycHuk — CHANGES_REQUESTED** [review]

**Comment by AlekSi** [comment]
https://github.com/percona/pmm/pull/570

**Review by BupycHuk — APPROVED** [review]

**Comment by AlekSi** [comment]
There are conflicts now

File: services/agents/mongodb.go [hunk 12]
```diff
@@ -43,10 +50,22 @@ func mongodbExporterConfig(service *models.Service, exporter *models.Agent, reda
...
+		fmt.Sprintf("HTTP_AUTH=pmm:%s", exporter.AgentID),
+	}
+
+	// Starting with PMM 2.10.0, we are shipping the new mongodb_exporter
+	if !pmmAgentVersion.Less(newMongoExporterPMMVersion) {
+		args = []string{
+			"--compatible-mode",
+			"--web.listen-address=:" + tdp.left + " .listen_port " + tdp.right,
+		}
+	}
```
Comment (AlekSi) [inline services/agents/mongodb.go:67]: I would change an order here.  Right now we have:
1. Flags for 2.9
2. Common environment variables.
3. Flags for 2.10.
4. Common flags.

That's a bit weird. How about moving environment variables before flags, and set flags in `if`? Like
```go
var args []string
if pmmAgentVersion.Less(newMongoExporterPMMVersion) {
  // flags for 2.9
} else {
  // flags for 2.10
}
```

File: services/agents/mongodb_test.go [hunk 13]
```diff
@@ -22,13 +22,16 @@ import (
...
 	"github.com/percona/pmm/api/inventorypb"
+	"github.com/percona/pmm/version"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
 	"github.com/percona/pmm-managed/models"
 )
 
 func TestMongodbExporterConfig(t *testing.T) {
+	pmmAgentVersion, _ := version.Parse("2.0.0")
```
Comment (AlekSi) [inline services/agents/mongodb_test.go:33]: MustParse

File: services/agents/mongodb_test.go [hunk 14]
```diff
@@ -65,13 +68,59 @@ func TestMongodbExporterConfig(t *testing.T) {
...
+
+	t.Run("EmptyUsername", func(t *testing.T) {
+		exporter.Username = nil
+		actual := mongodbExporterConfig(mongodb, exporter, exposeSecrets, pmmAgentVersion)
+		assert.Equal(t, "MONGODB_URI=mongodb://1.2.3.4:27017/?connectTimeoutMS=1000", actual.Env[0])
+	})
+}
+
+func TestNewMongodbExporterConfig(t *testing.T) {
+	pmmAgentVersion, _ := version.Parse("2.10.0")
```
Comment (AlekSi) [inline services/agents/mongodb_test.go:83]: MustParse

File: services/checks/checks_test.go [hunk 15]
```diff
@@ -421,20 +421,20 @@ func TestMustParseVersion(t *testing.T) {
...
-			mustParseVersion("")
+			version.MustParse("")
 		}
 		assert.Panics(t, f)
 	})
 
 	t.Run("invalid version", func(t *testing.T) {
 		f := func() {
-			mustParseVersion("invalid version")
+			version.MustParse("invalid version")
```
Comment (AlekSi) [inline services/checks/checks_test.go:437]: Well, here we test a function that is from the different repository and already tested. Probably we can remove it.

**Review by AlekSi — COMMENTED** [review]
A few minor comments, please act on them and update FB

File: services/checks/checks_test.go (hunk 15 above)
Comment (AlekSi) [reply services/checks/checks_test.go:437]: Actually, the whole TestMustParseVersion is not needed now :)

**Review by AlekSi — APPROVED** [review]
LGTM, but where is a feature build?

File: docker-compose.yml [hunk 16]
```diff
@@ -2,7 +2,7 @@
 version: '3.7'
 services:
   pmm-managed-server:
-    image: ${PMM_SERVER_IMAGE:-perconalab/pmm-server:dev-latest}
+    image: ${PMM_SERVER_IMAGE:-perconalab/pmm-server-fb:PR-1074-9585a6d}
```
Comment (AlekSi) [inline docker-compose.yml:5]: @BupycHuk Please revert that change right after release

Comment (AlekSi) [reply docker-compose.yml:5]: Or right now, as it looks like we still merge things

**Pull request merged**